Restrict named server servernames

manics opened the pull request:
See https://github.com/jupyterhub/jupyterhub/issues/5054
- Limits servernames to alphanumeric characters, and single `-`. See documentation in `slugs.py::safe_slug` or `howto/upgrading-v6.md` for details
  - Adds a separate display name with fewer restrictions since this is only for display purposes
- Existing servers can still be started
- [x] Add docs on how to migrate (probably increase the number of allowed named servers, start a new server, ask an admin to migrate the storage, and delete the old server? Or should we support starting legacy named servers indefinitely?)

minrk committed:
https://github.com/jupyterhub/kubespawner/blob/0020c0b39e4e8c9b072f91b2875af18b6c8f57ee/kubespawner/slugs.py

commit aa08a2d15520f9d9501d83209189afd997706b19
Author: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Date:   Mon Feb 3 23:59:20 2025 +0000

    [pre-commit.ci] auto fixes from pre-commit.com hooks

    for more information, see https://pre-commit.ci

commit eaf5cc3b5773df11dd404a6f6d5f1fd0553dcbda
Author: Erik Sundell <erik.i.sundell@gmail.com>
Date:   Fri Oct 18 10:29:58 2024 +0200

    Add escape_slug as a utility function alongside safe_slug

    By providing this, it becomes easier to test and compare slug changes.

commit fd0a507baf09ce43b704ba99d7781921af6c13d6
Author: Min RK <benjaminrk@gmail.com>
Date:   Wed Oct 2 11:23:25 2024 +0200

    fix some safe slug patterns

    - escape `-` in label pattern
    - don't accept `.` in object names, to allow satisfying strict RFC 1035/1123

commit dbc7682764e5538d3b0382e2de1003374f656e07
Author: Min RK <benjaminrk@gmail.com>
Date:   Wed Jun 26 11:07:54 2024 +0200

    sub '-' for any sequence of unsafe characters

    more recognizable with common separators like '@', '_', '.'

commit 143cc8d323db5e7dba445a8c83e5925f22059e5f
Author: Min RK <benjaminrk@gmail.com>
Date:   Mon Jun 17 16:12:12 2024 +0200

    add multi_slug mechanism for multi-word slugs (username--servername---hash)

commit 399b53a6ab944c1039e5a6454467d742e8724809
Author: Min RK <benjaminrk@gmail.com>
Date:   Mon Jun 17 11:17:33 2024 +0200

    let safe slug scheme live side-by-side with escape scheme

    - add user_server to template namespace
    - add `safe_` prefixed fields to template namespace using new 'safe' slug scheme
    - add pod_name, namespace, pvc_name to template namespace

commit d6cbf20caf850e430a1cad2a6fc0ea4566f29d8f
Author: Min RK <benjaminrk@gmail.com>
Date:   Wed Jun 12 14:42:01 2024 +0200

    toward compatible transition for safe slugs

    - safe scheme by default
    - preserve old scheme if found
    - persist pvc_name and attempt to handle lack of persistence from before
    - add handle_legacy_names and use_legacy_labels configuration for legacy labels

commit 68e00b92ffb1731e8ea0583e38e29ac6134c5a89
Author: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Date:   Wed Jun 12 07:59:05 2024 +0000

    [pre-commit.ci] auto fixes from pre-commit.com hooks

    for more information, see https://pre-commit.ci

commit 66462c5399a7a8e5c0a7a3a741c5209e546fdf1d
Author: Min RK <benjaminrk@gmail.com>
Date:   Thu Jun 8 12:51:16 2023 +0200

    add 'safe' slug scheme

    - always produces valid values
    - uses names, etc. as-is, when valid
    - uses stripped name--hash when needed

minrk committed:
https://github.com/jupyterhub/kubespawner/blob/7.0.0/tests/test_slugs.py

Extracted to jupyterhub/tests/test_slugs.py using git-filter-repo

toward compatible transition for safe slugs

- safe scheme by default
- preserve old scheme if found
- persist pvc_name and attempt to handle lack of persistence from before
- add handle_legacy_names and use_legacy_labels configuration for legacy labels

let safe slug scheme live side-by-side with escape scheme

- add user_server to template namespace
- add `safe_` prefixed fields to template namespace using new 'safe' slug scheme
- add pod_name, namespace, pvc_name to template namespace

sub '-' for any sequence of unsafe characters

more recognizable with common separators like '@', '_', '.'

document new template scheme and upgrade notes

consolidate references to template fields now that there's a doc for it

fix some safe slug patterns

- escape `-` in label pattern
- don't accept `.` in object names, to allow satisfying strict RFC 1035/1123

manics committed:
Restrict allowed names for named servers

manics commented on jupyterhub/utils.py:
```diff
@@ -956,6 +957,24 @@ def subdomain_hook_idna(name, domain, kind):
 
+# A URL safe identifier can be used in URL paths without escaping:
+# Unicode letters, digits, _, -
+_safe_unicode_identifier_re = re.compile(r'^[\w-]{1,63}$')
```
What's a good size limit? If we want something that can be converted to a DNS safe string and back then we need to take account of unicode being escaped. Or we could stick with the KubeSpawner approach of using a hash if the string is too long?

minrk replied:
if we do that, the 1123 check in kubespawner is [here](https://github.com/jupyterhub/kubespawner/blob/0020c0b39e4e8c9b072f91b2875af18b6c8f57ee/kubespawner/slugs.py#L65).

manics committed:
Add display_name to DB Spawners

manics committed:
Remove unneeded escapism from slugs.py

manics committed:
Set `safe_slug` `max_length=32`

manics committed:
Replace User.spawner SpawnerDict with explicit calls to get or create spawner

Rename `User.get_spawner` `get_or_create_spawner`

manics committed:
Add display_name in addition to server_name

manics committed:
Update named server tests

manics committed:
Bring back SpawnerDict

Have seperate `User.get_spawner` (requires orm_spawner to exist) and `User.get_or_create_spawner` (requires all parameters for creating a server to be passed, including `display_name`)

manics committed:
Need to save spawner in SpawnerDict

manics committed:
Relax safe_slug, allow alnum with optoinal - of any length

manics reviewed (commented):
I've had a go at adding a `display_name` field to the database, but it's ended up being more complicated than I anticipated and there some design decisions to make.

## Safe slugs

I copied `kubespawner/slugs.py` to `jupyterhub/slugs.py`, then realised it generates safe non-clashing names for combined username+servernames labels, e.g. it uses `---`. When naming a server we don't want this, since we may need to concatenate the servername later.

For now I went for something simpler, just `[alnum]` with an optional `-` in the middle. I haven't put a length limit yet, but if we set the length to `30` in `safe_slug` then in future we it would be fine to generate a DNS-safe label using `f"{safe_slug(username)}---{safe_slug(servername)}"`.

## SpawnerDict

This used to assume that if the ORM spawner doesn't exist you can just create it using `servername`. This is no longer the case, since `display_name` can be passed. To make things more explicit I've change dit so the SpawnerDict can be used to get a spawner if the ORM spawner exists, otherwise you must create the spawner using `User.get_or_create_spawner(server_name, display_name)`. I could make `display_name` default to `server_name` so that the SpawnerDict will still work as before, but I decided it's better to be explicit.

## Link between servername and display_name

The display name is actually completely independent of the server name. I've changed the Spawn UI so that the name entered on the form is treated as the display name, which is then converted to a servername using safe_slug. However this isn't strictly necessary, servername could be any safe_slug. If you use the API there's no requirement for `servername == safe_slug(display_name)`

minrk commented:
For SpawnerDict, I think we can make a `get_or_create` method. I don't think we need the defaultdict behavior, it may have been a little too much automatic logic anyway.

> Link between servername and display_name

Yeah, it's a totally free association. I'd probably leave the UI with a single display name as input and derive the true name via slugification, but let the API freely define the two.

manics committed:
Fix db upgrade error when running older upgrade script

This new scirpt upgrades the Spawner table which is referenced in an older upgrade script

manics committed:
Update db revisions f460b0348206_add_display_name_to_spawners.py

manics committed:
test_resume_spawners: spawners are no longer implicity created

manics committed:
Rebased: Revises afd65840b69e

manics committed:
Backend only validates server_name, frontend must create valid name

manics committed:
spawn page: use `?display_name=` query to match API

manics committed:
Don't treat `display_name` as an options form parameter

minrk commented:
That's a good point, though doing that means that we can't actually _assume_ names are safe anywhere, we have to keep everything that treats names as arbitrary strings everywhere. Can you think of an opt-in way to _guarantee_ that strict names are followed? Or perhaps a multi-stage deprecation, so it's a warning in jupyterhub 6, but error in jupyterhub 7?

manics commented:
I agree we don't want to support unsafe names indefinitely. Do you think we could add a flag to `jupyterhub upgrade-db` to auto-update names, have constant warnings for anyone who doesn't update them, and then make this a hard failure in JupyterHub 7?

minrk commented:
because upgrade-db runs precisely once, I think putting arguments on it is not likely to be super useful, because there's no way to re-run with different arguments. I think we'd need a hook somewhere to be able to do it after upgrade, either as a one-time hook or startup flag or something.

I think the hard failure in JupyterHub 7 can be that we don't allow starting the servers with mismatched names until they've been renamed, don't allow any operation on them other than delete & rename.

We could potentially do exactly that in JupyterHub 6: allow users to rename (and only allow safe names), but don't rename it for them, and allow admins to bulk auto-rename, if they want, just not at `upgrade-db` time. And prohibit _starting_ named servers with unsafe names, which means that the only APIs that need to deal with unsafe names are listing and renaming.

manics committed:
Don't forward display_name to spawn-pending

manics committed:
Allow disabling starting of existing invalid named servers

manics committed:
Check invalid named server can be deleted

manics committed:
allow_existing_invalid_named_servers=autorename

manics committed:
safe_slug: require non-empty string

manics committed:
Update browser test for spawning named server with form

manics committed:
Start adding docs on migrating named servers

https://jupyterhub.readthedocs.io/en/6.0.0/howto/upgrading-v6.html

manics committed:
WIP: thought more about slugs

manics committed:
Update slug tests

minrk commented on docs/source/howto/upgrading-v6.md:
```diff
@@ -0,0 +1,85 @@
+JupyterHub 6 imposes strict limits on the naming of named servers:
+
+- Between 1 and 32 characters
+- Consists of lowercase ASCII letters, digits, and zero or one hyphen (`-`)
```
why the limits on number of hyphens? And is this total or consecutive?

manics replied:
I think this is out of date, it should be single hyphens are allowed.

minrk commented on docs/source/howto/upgrading-v6.md:
```diff
@@ -0,0 +1,85 @@
+- Must start with a lowercase letter
+- Must end with a lowercase letter or digit
+
+A new field `display_name` means users can still provide an arbitrary name, separate from the server name.
```
```suggestion
A new field `display_name` means users can still provide an arbitrary name, separate from the server name used in APIs, URLs, etc.
```

minrk commented on docs/source/howto/upgrading-v6.md:
```diff
@@ -0,0 +1,85 @@
+### Web interface
+
+Users can name a named server as normal.
+This will be used as the display_name, and if this doesn't comply with the above restriction a server name will be automatically generated.
```
```suggestion
This will be used as the display_name, and if this doesn't comply with the above restriction, a safe server name will be automatically derived by stripping disallowed characters, truncating, as needed.
```

minrk reviewed (commented):
What do you need in order to finish this one? I think this is the main thing I would like to have finished to start preparing 6.0.

manics committed:
Apply suggestions from code review

Co-authored-by: Min RK <benjaminrk@gmail.com>

manics commented:
Let me rebase and check everything I've done, but I think all the main parts are in place (creating slugs, optionally updating names in the database).

manics committed:
rebase onto alembic version ff1860e38070

manics committed:
Fix tests

manics committed:
Validate display_name in User.get_or_create_spawner

manics committed:
upgrade doc: multiple single - allowed

manics commented on jupyterhub/app.py:
```diff
@@ -1367,6 +1369,26 @@ def named_server_limit_per_user_fn(handler):
         """,
     ).tag(config=True)
 
+    allow_existing_invalid_named_servers = Enum(
```
Naming is hard, happy to hear suggestions for alternatives!

manics commented on jupyterhub/app.py:
```diff
@@ -3287,6 +3309,9 @@ def init_tornado_settings(self):
             named_server_limit_per_user=self.named_server_limit_per_user,
+            allow_invalid_named_server_start=(
+                self.allow_existing_invalid_named_servers == "allow-start"
+            ),
```
I've named this tornado setting different;y from the Traitlet because the Application uses a multi-valued enum but the rest of the hub only cares about whether it's `allow-start` or not. Alternative is to pass the Traitlet through and deal with the enum later.

minrk replied:
I think that makes sense

manics commented on jupyterhub/app.py:
```diff
@@ -3338,6 +3363,67 @@ def init_eventlog(self):
+        ):
+            if not is_valid_safe_slug(spawner.name):
+                if spawner.display_name:
+                    # Is this excessively defensive?
```
Should I delete this? Should be impossible.

minrk replied:
I think it's fine since it's already here. "This shouldn't happen" checks do trigger once in a while, especially in unrealistic tests. And it will be rare, since post-upgrade this check will soon stop happening.

manics commented on jupyterhub/app.py:
```diff
@@ -3338,6 +3363,67 @@ def init_eventlog(self):
+                spawner.name = safe_name
+
+            self.db.commit()
+            # TODO: Do we need to handle errors and rollback?
```
Please check!

minrk replied:
Good checking! No, we don't generally need to deal with rollback during one-tie startup checks.

manics commented on jupyterhub/app.py:
```diff
@@ -3338,6 +3363,67 @@ def init_eventlog(self):
+            # TODO: Do we need to handle errors and rollback?
+
+            if not to_modify and not cant_modify:
+                self.log.info("All named servers have valid names")
```
Should this be logged at info or debug?

minrk replied:
probably debug

manics commented on share/jupyterhub/static/js/utils.js:
```diff
@@ -122,6 +122,65 @@ define(["jquery"], function ($) {
     modal.show();
   };
 
+  var safe_slug = function (name) {
```
Since this is (currently) only used in the front end for creating a servername we can use any algorithm, even a random string. For consistency I've reimplemented `jupyterhub/slugs.py::safe_slug` in JS, but the main downside is it adds a dependency on the `js-sha256` NPM package since SHA256 isn't available in standard JS.

manics committed:
Fix some formatting, remove unused code

manics committed:
Some fixes to utils.js::safe_slug

manics committed:
Add karma for testing, run `karma init`

AMD modules (requirejs) can't easily be tested with npm or jest

manics committed:
Add karma tests for utils.js::safe_slug

manics committed:
Run karma tests in CI

manics committed:
test-javascript.yml: zizmor ignores

manics committed:
Add 5.4.6 DB upgrade test

manics committed:
Add test for allow_existing_invalid_named_servers

manics commented on jupyterhub/slugs.py:
```diff
@@ -0,0 +1,274 @@
+"""Tools for generating slugs like object names and labels
```
Admittedly this is scope creep, but since we need a way to generate safe names I thought I'd write something generic that should hopefully work for escaping other names in future, including in external authenticators and spawners.

manics committed:
Fix comment in utils.js

manics commented on karma.conf.js:
```diff
@@ -0,0 +1,58 @@
+// Karma configuration
+// Generated on Sat May 30 2026 20:05:11 GMT+0100 (British Summer Time)
```
This file, and `test-main.js`, is the output of `karma init` with a few changes.

I looked at using Jest since that's what we're using in jsx, but couldn't get it to work with requirejs/AMD modules.

manics commented on package.json:
```diff
@@ -13,15 +13,21 @@
     "jsx:install-run": "npm install --prefix jsx && npm run --prefix jsx",
-    "jsx:run": "npm run --prefix jsx"
+    "jsx:run": "npm run --prefix jsx",
+    "test": "karma start"
```
This is potentially confusing to new contributors because this is a top-level `package.json` that runs tests, but we also have a completely independent `jsx/package.json` that also runs tests but is not triggered by this one.

manics reviewed (commented):
I think this is ready for review!

I've highlighted a few things/questions inline

One missing nice-to-have is a browser/playwright test to check a warning is shown for invalid servers on the home page:
<img width="1140" height="457" alt="Image" src="https://github.com/user-attachments/assets/249c67f9-fa16-473d-a338-c2277acffb3d" />

Preview of the 6.0.0 upgrade page: https://jupyterhub--5192.org.readthedocs.build/en/5192/howto/upgrading-v6.html

manics committed:
Validate server_name in post as well as get, normalise_unicode before validating, display_name from post must be string

manics commented:
If you want to test the handling of existing invalid servers without first installing JupyterHub 5 you can import this DB dump instead:
[jupyterhub5db.sql.txt](https://github.com/user-attachments/files/28445574/jupyterhub5db.sql.txt)

```sh
rm jupyterhub.sqlite
sqlite3 jupyterhub.sqlite '.read jupyterhub5db.sql.txt'

jupyterhub upgrade-db
```
`jupyterhub_config.py`:
```py
c.JupyterHub.spawner_class = "simple"
c.JupyterHub.authenticator_class = "dummy"
c.JupyterHub.allow_named_servers = True

## Set this to auto migrate:
#c.JupyterHub.allow_existing_invalid_named_servers = "autorename"
```

Run JupyterHub, login as user `demo`. You should have three named servers with invalid names.

minrk commented on jupyterhub/app.py:
```diff
@@ -3338,6 +3363,67 @@ def init_eventlog(self):
+                self.log.info("All named servers have valid names")
+
+        elif cant_modify or to_modify:
+            self.log.warning(
```
Great!

minrk reviewed (commented):
I think this is really great, and really appreciate the detailed docs.

My only point of disagreement right now is using the generic collision-safe hashing scheme on server names, which don't have any of the the challenges that motivated using it in kubespawner for resource names, because they are already namespaced per-user, so don't have the same security/collision consequences.

It doesn't seem right to me that `two-word` is not considered a safe server name. I really think single-hyphen should be allowed in safe names. I understand the reasoning that it should be valid to combine two slugs with e.g. `--`, but server names do not have the same collision issues that e.g. usernames have, so I don't think there's a reason to use the hashing scheme in server names.

Adding hashes also increases the likelihood of needing _further_ truncation and hashing due to increasing the length of simple strings.

I think if we make the collision-avoidance hash _optional_ and opt-out for server names, that would work for me

Then we can have:

```python
safe_slug("a-b") # a-b-c8687a08
safe_slug("a-b", avoid_collisions=False) # 'a-b'
```

and use the latter for server names, and the former where collisions must be avoided.

manics committed:
safe_slug: optionally disable hash, try and keep more chars

manics committed:
Frontend: don't hash slugs

manics committed:
Don't hash autorenamed servers

Unless they clash

manics committed:
Small cleanups

manics commented:
I changed safe_slug as you suggested and removed the hash from servernames form back and front ends, the exception is if there's an actual clash in the autorename process.

One problem is I haven't done any checks for collision on the front end because there's no easy way to get the list of servernames without making an additional API request. If the entered name for a new server corresponds to a safe_name that clashes with an existing server the user is taken to that existing server. The behaviour of the spawn request is technically unchanged from before this PR, but it's now a problem because we're silently converting the user supplied name to the safe_name.

minrk commented:
That makes sense, I don't think the frontend should check for collisions. This reveals a problem in our API that creating a named server and starting it for the first time are the same, where it should actually be two steps:

1. create the server (this should 409 on name collision), _then_
2. start the new server

I think we should land this and then look into improving that sequence by adding a new API endpoint that only creates the server instead of starting it.

manics commented:
Thanks, I think this is ready then! Be aware that if there are problems after merging I won't be able to do any follow-up work for the next 3 weeks.

minrk commented:
Thanks for all your work on this!
